Expand NT subclass to support SAM

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111078
* __->__ #109123

This PR contains the changes needed to support using the NT jagged subclass within SAM. Note that a NT with multiple ragged dims is still required at the extremes for inputs / outputs, but the internal computation generally involves a single ragged dim, making the jagged layout usable.

jbschlosser committed:
Expand NT subclass to support SAM

[ghstack-poisoned]

jbschlosser committed:
Update on "Expand NT subclass to support SAM"


This PR contains the changes needed to support using the NT jagged subclass within SAM. Note that a NT with multiple ragged dims is still required at the extremes for inputs / outputs, but the internal computation generally involves a single ragged dim, making the jagged layout usable.

[ghstack-poisoned]

jbschlosser committed:
Update on "Expand NT subclass to support SAM"


This PR contains the changes needed to support using the NT jagged subclass within SAM. Note that a NT with multiple ragged dims is still required at the extremes for inputs / outputs, but the internal computation generally involves a single ragged dim, making the jagged layout usable.

[ghstack-poisoned]

jbschlosser committed:
Update on "Expand NT subclass to support SAM"


This PR contains the changes needed to support using the NT jagged subclass within SAM. Note that a NT with multiple ragged dims is still required at the extremes for inputs / outputs, but the internal computation generally involves a single ragged dim, making the jagged layout usable.

[ghstack-poisoned]

jbschlosser committed:
Update on "Expand NT subclass to support SAM"


This PR contains the changes needed to support using the NT jagged subclass within SAM. Note that a NT with multiple ragged dims is still required at the extremes for inputs / outputs, but the internal computation generally involves a single ragged dim, making the jagged layout usable.

[ghstack-poisoned]

jbschlosser committed:
Update on "Expand NT subclass to support SAM"


This PR contains the changes needed to support using the NT jagged subclass within SAM. Note that a NT with multiple ragged dims is still required at the extremes for inputs / outputs, but the internal computation generally involves a single ragged dim, making the jagged layout usable.

[ghstack-poisoned]

jbschlosser committed:
Update on "Expand NT subclass to support SAM"


This PR contains the changes needed to support using the NT jagged subclass within SAM. Note that a NT with multiple ragged dims is still required at the extremes for inputs / outputs, but the internal computation generally involves a single ragged dim, making the jagged layout usable.

[ghstack-poisoned]

jbschlosser committed:
Update on "Expand NT subclass to support SAM"


This PR contains the changes needed to support using the NT jagged subclass within SAM. Note that a NT with multiple ragged dims is still required at the extremes for inputs / outputs, but the internal computation generally involves a single ragged dim, making the jagged layout usable.

[ghstack-poisoned]

jbschlosser committed:
Update on "Expand NT subclass to support SAM"


This PR contains the changes needed to support using the NT jagged subclass within SAM. Note that a NT with multiple ragged dims is still required at the extremes for inputs / outputs, but the internal computation generally involves a single ragged dim, making the jagged layout usable.

[ghstack-poisoned]

jbschlosser committed:
Update on "Expand NT subclass to support SAM"


This PR contains the changes needed to support using the NT jagged subclass within SAM. Note that a NT with multiple ragged dims is still required at the extremes for inputs / outputs, but the internal computation generally involves a single ragged dim, making the jagged layout usable.

[ghstack-poisoned]

soulitzer commented on torch/nested/_internal/ops.py:
```diff
@@ -172,19 +192,226 @@ def tensor_attr_unsupported_getter(func, *args, **kwargs):
+    new_values = torch.mm(inp._values, weight)
+    if bias is not None:
+        new_values += bias
+    return NestedTensor(new_values, inp._offsets)
```
Unfortunately, for the symbolic singleton symint to be propagated, we should also need to thread the `inp._ragged_size` around. For now I'm using `extract_kwargs` as a utility to propagate all necessary metadata.

```suggestion
    return NestedTensor(new_values, **extract_kwargs(inp))
```

soulitzer replied:
Oh yeah this is probably rebase shenanigans, thanks for doing the rebase :)

soulitzer commented on torch/_prims_common/__init__.py:
```diff
@@ -1494,11 +1494,17 @@ def make_contiguous_strides_for(
 
+    def _ensure_at_least(a, thresh):
+        # sym_max() isn't supported for SingletonSymNode
+        if hasattr(a, "node") and a.node.singleton_int() is not None:
```
nit:
```suggestion
        if isinstance(a, torch.SymInt) and a.node.singleton_int() is not None:
```
hmmm interesting why its still going to the other path, you might want to check if threshold is also a singleton int?

jbschlosser replied:
Discussed offline: we need to check for SymInt wrapping a SingletonSymNode as well (this happens for the fake-ified case)

jbschlosser committed:
Update on "Expand NT subclass to support SAM"


This PR contains the changes needed to support using the NT jagged subclass within SAM. Note that a NT with multiple ragged dims is still required at the extremes for inputs / outputs, but the internal computation generally involves a single ragged dim, making the jagged layout usable.

[ghstack-poisoned]
